# EDGES
## Overview
#### Figure: EDGES Framework 
EDGES is a spatially constrained non-negative matrix factorization (NMF) framework designed for gene expression denoising and prediction in spatial transcriptomic (ST) data.![EDGES overview](https://i.imgur.com/SpWkY1M.png)
#  Tutorials
This tutorial provides detailed instructions for running EDGES in both MATLAB and Python.

## Inputs
**X**: `shared genes by cells` matrix of ST data.\
**Y**:  `shared genes by cells`  matrix of scRNA-seq data.\
**Z**: `unique genes by cells` matrix of scRNA-seq data.\
**L1** :  `cells by cells` matrix. A precomputed normalized Laplacian matrix representing the spatial network structure.\
**lambda1**: Hyper-parameter for the spatial constraint.\
**lambda2**: Hyper-parameter for the sparse term.\
**theta1**: Hyper-parameter for the decomposition of ST data.\
**theta2**: Hyper-parameter for the decomposition of unique scRNA-seq data.\
**tol**: Stopping condition.\
**d**: Dimensionality of the low-dimensional space.\
**iterMax**: Maximum number of iterations.
## Outputs:
**W1**: The low-dimensional representation of shared genes.\
**W2**: The low-dimensional representation of unique genes.\
**H1**: The low-dimensional representation of ST cells.\
**H2**: The low-dimensional representation of scRNA-seq cells.\
**p1**: Denoised ST data after denoising.\
**p2**: Predicted ST data generated by EDGES.
## Data preprocess
The raw ST data retains only the genes shared with scRNA-seq data, and the scRNA-seq data is divided into shared and unique scRNA-seq data based on these shared genes.
## How to Run (Matlab)
### Version
Matlab 2020a
### How to run
You can execute the model as follows:
```matlab
​[W1, W2, H1, H2, p1, p2] = EDGES(X, Y, Z, L1, lambda1, lambda2, theta1, theta2, tol, d, iterMax);
```
## How to Run (Python)
### Version
```bash
python >= 3.8

# Install dependencies
pip install numpy scipy pandas scikit-learn
```
### How to run
You can execute the model as follows:
```python
from EDGES import EDGES
# Load or simulate your matrices:
X = "stdata.csv"   # ST data
Y = "scdata_sahred.csv"   # shared scRNA-seq data
Z = "scdata_unique.csv"   # unique scRNA-seq data
L1 = "laplacian.csv"  # normalized graph Laplacian
# Set hyperparameters
lambda1 = 1e-5
lambda2 = 1e1
theta1 = 1e-1
theta2 = 1e-4
tol = 1e-7
d = 20
iterMax = 500

# Run EDGES
W1, W2, H1, H2, p1, p2 = EDGES(X, Y, Z, L1, lambda1, lambda2, theta1, theta2, tol, d, iterMax)
```
You can also run the Python version of EDGES via demo.ipynb.
## Reference
Jinyue Zhao, Jiating Yu, Yuqing Cao, Fan Yuan, Ling-Yun Wu, and Duanchen Sun. "Enhancing Spatial Transcriptomics via Spatially Constrained Matrix Decomposition with EDGES."


